### PR TITLE
server,cks: check if vm is cks node during vm destroy

### LINKED ISSUE #8902: CKS Cluster goes into alert state when a node is expunged
ISSUE TYPE

Improvement request/BUG

COMPONENT NAME

Component: CKS

CLOUDSTACK VERSION

Cloudstack version 4.18 4.19

SUMMARY

CKS Cluster goes into an alert state when a vm part of the cluster is  destoryed/expunged

Steps to reproduce the issue

1. Create a cks cluster with a worker size let's say 2
2. Wait till cks cluster is in running state
3.  Expunge one of the worker node vm
4. The cks cluster goes into alert state

Expected behaviour 

Cloudstack should throw an error popup message when user tries to delete a vm which is part of a cks cluster.

"Expunge is not allowed as its part of a cks cluster."

This could be the same behaviour when we try to delete a vpc which has active tier in it 

Actual behaviour 

Currently, Cloudstack allows to deletion a vm which is part of a cks cluster

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes: #8902

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

```
(local) 🐱 > list kubernetesclusters filter=id,virtualmachines
{
  "count": 1,
  "kubernetescluster": [
    {
      "id": "a2cd4198-55be-43da-9a3a-1a05ebb9745d",
      "virtualmachines": [
        {
          "account": "ACSUser",
          "affinitygroup": [],
          "created": "2024-06-05T07:51:22+0000",
          "details": {},
          "displayname": "test1-control-18fe761907b",
          "domain": "ROOT",
          "domainid": "8da34656-08a3-11ef-9779-1e00a7000c5f",
          "guestosid": "8da8cd0e-08a3-11ef-9779-1e00a7000c5f",
          "haenable": false,
          "hasannotations": false,
          "hostcontrolstate": "Enabled",
          "hypervisor": "KVM",
          "id": "77919cbe-e0de-48c9-aa97-fe789f7ccbab",
          "isdynamicallyscalable": false,
          "lastupdated": "2024-06-05T07:51:34+0000",
          "name": "test1-control-18fe761907b",
          "nic": [
            {
              "broadcasturi": "vlan://2007",
              "deviceid": "0",
              "extradhcpoption": [],
              "gateway": "10.1.1.1",
              "id": "0d04047c-7888-4c8f-8a85-da2d1d0c7aac",
              "ipaddress": "10.1.1.130",
              "isdefault": true,
              "isolationuri": "vlan://2007",
              "macaddress": "02:01:00:d7:00:02",
              "netmask": "255.255.255.0",
              "networkid": "e8e2095e-6575-4639-8dbd-411e14b6753f",
              "networkname": "test1-network",
              "secondaryip": [],
              "traffictype": "Guest",
              "type": "Isolated"
            }
          ],
          "osdisplayname": "Debian GNU/Linux 5.0 (64-bit)",
          "ostypeid": "8da8cd0e-08a3-11ef-9779-1e00a7000c5f",
          "pooltype": "NetworkFilesystem",
          "readonlydetails": "dataDiskController, rootDiskController",
          "receivedbytes": 0,
          "securitygroup": [],
          "sentbytes": 0,
          "state": "Running",
          "tags": [],
          "userid": "a709d3e3-3254-40b2-bdc7-79c269e84ac7",
          "username": "user",
          "zoneid": "31c7d02e-7734-4bef-ab3c-e48a244025ce",
          "zonename": "pr8601-t6927-kvm-centos7"
        },
        {
          "account": "ACSUser",
          "affinitygroup": [],
          "created": "2024-06-05T07:51:34+0000",
          "details": {
            "deployvm": "true"
          },
          "displayname": "test1-node-18fe761c02e",
          "domain": "ROOT",
          "domainid": "8da34656-08a3-11ef-9779-1e00a7000c5f",
          "guestosid": "8da8cd0e-08a3-11ef-9779-1e00a7000c5f",
          "haenable": false,
          "hasannotations": false,
          "hypervisor": "KVM",
          "id": "8fa9e120-cc14-4973-95f9-bd68776aae16",
          "isdynamicallyscalable": false,
          "lastupdated": "2024-06-05T07:51:37+0000",
          "name": "test1-node-18fe761c02e",
          "nic": [
            {
              "deviceid": "0",
              "extradhcpoption": [],
              "gateway": "10.1.1.1",
              "id": "9e98a986-6d77-4083-aadb-8a447635d1bd",
              "ipaddress": "10.1.1.112",
              "isdefault": true,
              "macaddress": "02:01:00:d7:00:03",
              "netmask": "255.255.255.0",
              "networkid": "e8e2095e-6575-4639-8dbd-411e14b6753f",
              "networkname": "test1-network",
              "secondaryip": [],
              "traffictype": "Guest",
              "type": "Isolated"
            }
          ],
          "osdisplayname": "Debian GNU/Linux 5.0 (64-bit)",
          "ostypeid": "8da8cd0e-08a3-11ef-9779-1e00a7000c5f",
          "pooltype": "NetworkFilesystem",
          "readonlydetails": "dataDiskController, rootDiskController",
          "receivedbytes": 0,
          "securitygroup": [],
          "sentbytes": 0,
          "state": "Stopped",
          "tags": [],
          "userid": "a709d3e3-3254-40b2-bdc7-79c269e84ac7",
          "username": "user",
          "zoneid": "31c7d02e-7734-4bef-ab3c-e48a244025ce",
          "zonename": "pr8601-t6927-kvm-centos7"
        }
      ]
    }
  ]
}
(local) 🐱 > destroy virtualmachine id=77919cbe-e0de-48c9-aa97-fe789f7ccbab expunge=true
{
  "account": "admin",
  "accountid": "aaa96ed7-08a3-11ef-9779-1e00a7000c5f",
  "cmd": "org.apache.cloudstack.api.command.admin.vm.DestroyVMCmdByAdmin",
  "completed": "2024-06-05T08:38:15+0000",
  "created": "2024-06-05T08:38:15+0000",
  "domainid": "8da34656-08a3-11ef-9779-1e00a7000c5f",
  "domainpath": "ROOT",
  "jobid": "5370d57d-0709-4463-b5cc-21a8fa34f3a0",
  "jobinstanceid": "77919cbe-e0de-48c9-aa97-fe789f7ccbab",
  "jobinstancetype": "VirtualMachine",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 530,
    "errortext": "Instance is a part of a Kubernetes cluster: test1. Use Instance delete option from Kubernetes cluster details or scale API for Kubernetes clusters with 'nodeids' to destroy the instance."
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "aaaa743f-08a3-11ef-9779-1e00a7000c5f"
}
🙈 Error: async API failed for job 5370d57d-0709-4463-b5cc-21a8fa34f3a0
```

### CODE REVIEW

**Comment by weizhouapache** [comment]
@shwstppr 
I am not sure if we should disallow the operation.

would it be a valid use case for users to
- drain the node in CKS
- remove the node vm from cloudstack
- scale the cks cluster to get a new node

for your information, in Autoscale VM group, users can destroy a specific VM when the vm group is disabled.
the corresponding reference/map in database and load balancer rules are also removed.
of course, cks cluster is totally different from AS vm group.

**Review by sureshanaparti — APPROVED** [review]
clgtm

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java [hunk 1]
```diff
@@ -16,26 +16,48 @@
...
+    public void checkVmCanBeDestroyed(UserVm userVm) {
+        if (!UserVmManager.CKS_NODE.equals(userVm.getUserVmType())) {
+            return;
+        }
+        KubernetesClusterVmMapVO vmMapVO = kubernetesClusterVmMapDao.findByVmId(userVm.getId());
+        if (vmMapVO == null) {
+            return;
+        }
+        logger.error(String.format("VM ID: %s is a part of Kubernetes cluster ID: %d", userVm.getId(), vmMapVO.getClusterId()));
+        throw new CloudRuntimeException("Instance is a part of a Kubernetes cluster");
```
Comment (sureshanaparti) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:58]: specify the Kubernetes cluster id/name in the msg passed to the response?

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java [hunk 2]
```diff
@@ -16,26 +16,48 @@
...
+    @Inject
+    private KubernetesClusterVmMapDao kubernetesClusterVmMapDao;
 
     @Override
     public ControlledEntity findByUuid(String uuid) {
         return kubernetesClusterDao.findByUuid(uuid);
     }
 
+    @Override
+    public void checkVmCanBeDestroyed(UserVm userVm) {
```
Comment (sureshanaparti) [inline plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: can destroy allowed with force option?

**Comment by shwstppr** [comment]
> @shwstppr I am not sure if we should disallow the operation.
> 
> would it be a valid use case for users to
> 
>     * drain the node in CKS
> 
>     * remove the node vm from cloudstack
> 
>     * scale the cks cluster to get a new node
> 
> 
> for your information, in Autoscale VM group, users can destroy a specific VM when the vm group is disabled. the corresponding reference/map in database and load balancer rules are also removed. of course, cks cluster is totally different from AS vm group.

@weizhouapache understood. Though destroying a valid cks node is also allowed right now.
@Pearl1594 @nvazquez do we plan to add an API to remove a VM/node from CKS cluster?
If there is, then the case mentioned by Wei should get handled? User can:
- drain node
- remove it from k8s cluster
- destroy the vm

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java (hunk 1 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:58]: will add

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java (hunk 2 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: do you mean adding a new API param as there is no force param in destroyVirtualMachine?

Comment (sureshanaparti) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: yes, with force param in destroy call.

**Review by rohityadavcloud — APPROVED** [review]

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java (hunk 2 above)
Comment (vishesh92) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: +1

**Comment by shwstppr** [comment]
@weizhouapache on the special case discussed earlier, @Pearl1594 made me aware that we already have an option with scaleKubernetesCluster API to delete a specific node using `nodeids` param. We also show a delete action in the UI so the use-case that you mentioned should get covered by that,
image

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java (hunk 2 above)
Comment (shwstppr) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: @sureshanaparti @vishesh92 I'm not completely in favour of adding the parameter. Deleting specific node of a k8s cluster is already covered by scaleKubernetesCluster API
see https://github.com/apache/cloudstack/pull/9057#issuecomment-2102512565

Do we have any other specific case for which we should allow the operation?

**Comment by weizhouapache** [comment]
> @weizhouapache on the special case discussed earlier, @Pearl1594 made me aware that we already have an option with scaleKubernetesCluster API to delete a specific node using `nodeids` param. We also show a delete action in the UI so the use-case that you mentioned should get covered by that,
> image
> 

That is great, thanks @shwstppr 

Overall this pr looks good to me.

Can we add the information above in the error message, if user tries to delete a cks node?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
Added test API calls in the PR description

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Review by DaanHoogland — APPROVED** [review]
clgtm and 
image

File: plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java (hunk 2 above)
Comment (vishesh92) [reply plugins/integrations/kubernetes-service/src/main/java/com/cloud/kubernetes/cluster/KubernetesClusterHelperImpl.java:49]: No other specific use case from my side.

**Pull request merged**